### ISSUE TITLE
Custom objective function, where to put the derivative

### ISSUE THREAD

**Issue by ChuanWang90** [keras-team/keras#8356]
I am trying to modify a bit the loss function of my convent and I have some questions from the implementation side.

I already know how to create a custom loss function in Keras, and how to call it. But I still do not have clear where to include the derivative of the function.

Let's say that my new loss function is:

Loss = cross-entropy + f(x)

where f(x) = x**2.

Where should I include f'(x)=2x so that it is used in the back-prop step? Does Keras automatically do that? Or should I define this explicitly in some part?

Thanks for any hint on this, since I do not know how to do it.

Chuan.

**Comment by ozabluda** [comment]
Keras have symbolic auto-differentiation, which will automatically and optimally do it for you.

**Comment by ChuanWang90** [comment]
Thanks for your answer @ozabluda I thought there should be something like that but it wasn't very clear.

**Comment by rjpg** [comment]
hello,

I want to buids a custom loss function that is equal to categorical_crossentropy  but in some predictions I want to devide by 2.

for example If I have 5 classes and the predicted is A but the real one is B I want to divide the result of categorical_crossentropy by 2. 

If I use some complex python function with cycles and "if"'s I cant see how keras will process the derivative ...

I explain better here :
https://stats.stackexchange.com/questions/389450/changing-the-categorical-crossentropy-loss-function-in-keras


Thanks !
